pkg/backend/blobstate/storage/azure/bucket.go

```diff
@@ -0,0 +1,173 @@
+type Bucket struct {
+	url *azblob.ContainerURL
+}
+
+// AccountName returns the buckets associated Azure Blob Storage account name
+func (b *Bucket) AccountName() string {
+	u := b.url.URL()
+	return extractAccountNameFromURL(&u)
+}
+
+func extractAccountNameFromURL(u *url.URL) string {
+	return strings.Split(u.Hostname(), ".")[0]
+}
+
+// NewBucket creates a new Bucket instance
+func NewBucket(url, accountKey string) (storage.Bucket, error) {
+
+	contURL, err := newContainerURL(url, accountKey)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = contURL.Create(context.Background(), azblob.Metadata{}, azblob.PublicAccessNone)
+	if err != nil {
+		if serr, ok := err.(azblob.StorageError); ok {
+			if serr.ServiceCode() != azblob.ServiceCodeContainerAlreadyExists {
+				return nil, err
+			}
+		}
+	}
+
+	bucket := Bucket{
+		url: contURL,
+	}
+
+	return &bucket, nil
+}
+
+func newContainerURL(cloudURL, accountKey string) (*azblob.ContainerURL, error) {
+	URL, err := url.Parse(strings.Replace(cloudURL, URLPrefix, "https://", 1))
+	if err != nil {
+		return nil, err
+	}
+
+	accountName := extractAccountNameFromURL(URL)
+	creds := azblob.NewSharedKeyCredential(accountName, accountKey)
+	pipe := azblob.NewPipeline(creds, azblob.PipelineOptions{
+		Retry: azblob.RetryOptions{
+			Policy:   azblob.RetryPolicyExponential,
+			MaxTries: 3,
+		},
+	})
+
+	cURL := azblob.NewContainerURL(*URL, pipe)
+	return &cURL, nil
+}
+
+// ListBlobs will list all blobs under a given prefix. This method will handle paging of
+// responses and create an aggregated array of blob paths. We don't anticipate loading
+// enough blobs to cause any significant memory pressure in Pulumi.
+func (b *Bucket) ListBlobs(ctx context.Context, prefix string) ([]string, error) {
+	var blobs []string
+
+	// Page through all the blobs in the container and build a list of blob names
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		listBlob, err := b.url.ListBlobs(ctx, marker, azblob.ListBlobsOptions{
+			Prefix: prefix,
+		})
+		if err != nil {
+			return nil, err
+		}
+		marker = listBlob.NextMarker
+		for _, blobInfo := range listBlob.Blobs.Blob {
+			blobs = append(blobs, blobInfo.Name)
+		}
+	}
+	return blobs, nil
+}
+
+// WriteBlob will create a new block blob
+func (b *Bucket) WriteBlob(ctx context.Context, path string, bytes []byte) error {
+	blobURL := b.url.NewBlockBlobURL(path)
+	numCores := uint16(runtime.NumCPU())
+	_, err := azblob.UploadBufferToBlockBlob(ctx, bytes, blobURL, azblob.UploadToBlockBlobOptions{
+		Parallelism: numCores,
+	})
+	return err
+}
+
+// ReadBlob will read the specified blob's contents
+func (b *Bucket) ReadBlob(ctx context.Context, path string) ([]byte, error) {
+	blobURL := b.url.NewBlockBlobURL(path)
+	res, err := blobURL.GetBlob(ctx, azblob.BlobRange{}, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		return nil, err
+	}
+	blob, err := ioutil.ReadAll(res.Body())
+	if err != nil {
```

Comment:
Do we need to close the body after reading it?

Reply:
Yes, good catch - I'll add a deferred close